### PR TITLE
[string] Shrink storage class sizes

### PR DESCRIPTION
**Pull request by milseman**
[string] Shrink storage class sizes.

* Don't allocate breadrumbs pointer if under threshold
* Increase breadrumbs threshold
* Linear 16-byte bucketing until 128 bytes, malloc_size after
* Allow cap less than _SmallString.capacity (bridging non-ASCII)

This change decreases the amount of heap usage for moderate-length
strings (< 64 UTF-8 code units in length) and increases the amount of
spare code unit capacity available (less growth needed).

Average improvements for moderate-length strings:

* 64-bit: on average, 8 bytes saved and 4 bytes of extra capacity
* 32-bit: on average, 4 bytes saved and 6 bytes of extra capacity

Additionally, on 32-bit, large-length strings also gain an average of
6 bytes of extra spare capacity.

Details:

On 64-bit, half of moderate-length allocations will save 16 bytes
while the other half get an extra 8 bytes of spare capacity.

On 32-bit, a quarter of moderate-length allocations will save 16
bytes, and the rest get an extra 4 bytes of spare
capacity. Additionally, 32-bit string's storage class now claims its
full allocation, which is its birthright. Prior to this change, we'd
have on average 1.5 bytes of spare capacity, and now we have 7.5 bytes
of spare capacity.

Breadcrumbs threshold is increased from the super-conservative 32 to
the pretty-conservative 64. Some speed improvements are incorporated
in this change, but more are in flight. Even without those eventual
improvements, this is a worthwhile change (ASCII is still fast-pathed
and irrelevant to breadcrumbing).

For a complex real-world workload, this amounts to around a 5%
improvement to transient heap usage due to all strings and a 4%
improvement to peak heap usage due to all strings. For moderate-length
strings specifically, this gives around 11% improvement to both.

rdar://problem/59685567

### CODE REVIEW

**Comment by milseman** [comment]
@swift-ci please test

**Comment by milseman** [comment]
@swift-ci please benchmark

**Comment by milseman** [comment]
@swift-ci please test

File: stdlib/public/core/LegacyABI.swift [hunk 1]
```diff
@@ -79,3 +79,11 @@ extension String {
     range: Range<String.Index>
   ) { fatalError() }
 }
+
+extension String.UTF16View {
+  // Swift 5.x: This was accidentally shipped as inlinable, but was never used
```
Comment (Catfish-Man) [inline stdlib/public/core/LegacyABI.swift]: :(

File: stdlib/public/core/StringStorage.swift [hunk 2]
```diff
@@ -49,42 +241,38 @@ final internal class __StringStorage
...
-#else
-  // The capacity of our allocation. Note that this includes the nul-terminator,
-  // which is not available for overriding.
-  internal var _realCapacityAndFlags: UInt64
-  internal var _countAndFlags: _StringObject.CountAndFlags
 
   @inline(__always)
-  internal var count: Int { return _countAndFlags.count }
+  fileprivate var _capacityAndFlags: _CapacityAndFlags {
+    // TODO: ...
```
Comment (Catfish-Man) [inline stdlib/public/core/StringStorage.swift]: ?

File: stdlib/public/core/StringStorage.swift [hunk 3]
```diff
@@ -93,102 +281,88 @@ final internal class __StringStorage
...
+    // We can check layout invariants, but our code units have not yet been
+    // initialized so we can't verify e.g. ASCII-ness
+    storage._invariantCheck(initialized: false)
     return storage
   }
 
-  @_effects(releasenone)
-  private static func create(
-    capacity: Int, countAndFlags: CountAndFlags
+  // DO NOT PUSH: For stats gathering only...
```
Comment (Catfish-Man) [inline stdlib/public/core/StringStorage.swift]: Just making sure you noticed this :)

File: stdlib/public/core/StringStorage.swift [hunk 4]
```diff
@@ -93,102 +281,88 @@ final internal class __StringStorage
...
-    _internalInvariant(capacity >= countAndFlags.count)
-
-    let realCapacity = determineCodeUnitCapacity(capacity)
-    _internalInvariant(realCapacity > capacity)
-    return __StringStorage.create(
-      realCodeUnitCapacity: realCapacity, countAndFlags: countAndFlags)
+    __StringStorage.create(
+      codeUnitCapacity: capacity,
+      countAndFlags: _CountAndFlags(mortalCount: count, isASCII: false))
+    // NOTE: content is garbage and not necessarily validy endcoded. Please
```
Comment (Catfish-Man) [inline stdlib/public/core/StringStorage.swift]: s/endcoded/encoded

File: stdlib/public/core/StringStorage.swift [hunk 5]
```diff
@@ -93,102 +281,88 @@ final internal class __StringStorage
...
     let buffer = UnsafeMutableBufferPointer(start: storage.mutableStart,
                                             count: capacity)
     let count = try initializer(buffer)
 
-    let countAndFlags = CountAndFlags(mortalCount: count, isASCII: false)
+    let countAndFlags = _CountAndFlags(mortalCount: count, isASCII: false)
     #if arch(i386) || arch(arm) || arch(wasm32)
     storage._count = countAndFlags.count
-    storage._flags = countAndFlags.flags
+    storage._countFlags = countAndFlags.flags
```
Comment (Catfish-Man) [inline stdlib/public/core/StringStorage.swift]: having count, countFlags, and countAndFlags on the same thing is a bit confusing heh

File: stdlib/public/core/StringStorage.swift [hunk 6]
```diff
@@ -522,3 +717,40 @@ extension __SharedStringStorage {
...
+  }
+
+  @inline(never) // slow-path
+  @_effects(releasenone)
+  internal func populateBreadcrumbs(
+    _ mutPtr: UnsafeMutablePointer<_StringBreadcrumbs?>
+  ) {
+    // Thread-safe compare-and-swap
+    let crumbs = _StringBreadcrumbs(String(self))
+    _stdlib_atomicInitializeARCRef(
```
Comment (Catfish-Man) [inline stdlib/public/core/StringStorage.swift]: huh TIL about this

**Review by Catfish-Man — APPROVED** [review]

**Comment by milseman** [comment]
@swift-ci please test

**Pull request merged**